security/pkg/pki/ca/selfsignedcarootcertrotator.go

```diff
@@ -0,0 +1,210 @@
+			certEncoded := base64.StdEncoding.EncodeToString(rotator.ca.GetCAKeyCertBundle().GetRootCertPem())
+			if err = rotator.configMapController.InsertCATLSRootCertWithRetry(
+				certEncoded, rotator.config.retryInterval, 30*time.Second); err != nil {
+				rootCertRotatorLog.Errorf("Failed to write self-signed Citadel's root cert "+
+					"to configmap (%s). Node agents will not be able to connect.",
+					err.Error())
+			}
+			rootCertRotatorLog.Info("Root cert is updated into configmap.")
+		}
+		return
+	}
+
+	rootCertRotatorLog.Infof("Refresh root certificate, root cert is about to expire: %s", err.Error())
+	options := util.CertOptions{
+		TTL:           rotator.config.caCertTTL,
+		SignerPrivPem: caSecret.Data[caPrivateKeyID],
+		Org:           rotator.config.org,
+		IsCA:          true,
+		IsSelfSigned:  true,
+		RSAKeySize:    caKeySize,
+		IsDualUse:     rotator.config.dualUse,
+	}
+	pemCert, pemKey, ckErr := util.GenRootCertFromExistingKey(options)
+	if ckErr != nil {
+		rootCertRotatorLog.Errorf("unable to generate CA cert and key for self-signed CA: %s", ckErr.Error())
+		return
+	}
+
+	oldCaCert := caSecret.Data[caCertID]
+	oldCaPrivateKey := caSecret.Data[caPrivateKeyID]
+	caSecret.Data[caCertID] = pemCert
+	caSecret.Data[caPrivateKeyID] = pemKey
+	if err = rotator.caSecretController.UpdateCASecretWithRetry(caSecret, rotator.config.retryInterval, 30*time.Second); err != nil {
+		rootCertRotatorLog.Errorf("Failed to write secret to CA secret (error: %s). "+
+			"Abort new root certificate.", err.Error())
+		return
+	}
+	rootCertRotatorLog.Infof("A new self-generated root certificate is written into secret: %v", string(pemCert))
+
+	if err := rotator.ca.GetCAKeyCertBundle().VerifyAndSetAll(pemCert, pemKey, nil, pemCert); err != nil {
+		rootCertRotatorLog.Errorf("Failed to create CA KeyCertBundle (%s)", err.Error())
+		return
+	}
+	rootCertRotatorLog.Infof("Updated CA KeyCertBundle using existing public key: %v", string(pemCert))
+
+	certEncoded := base64.StdEncoding.EncodeToString(rotator.ca.GetCAKeyCertBundle().GetRootCertPem())
+	if err = rotator.configMapController.InsertCATLSRootCertWithRetry(
+		certEncoded, rotator.config.retryInterval, 30*time.Second); err != nil {
+		rootCertRotatorLog.Errorf("Failed to write self-signed Citadel's root cert "+
+			"to configmap (%s). Rollback root cert.", err.Error())
+		caSecret.Data[caCertID] = oldCaCert
+		caSecret.Data[caPrivateKeyID] = oldCaPrivateKey
+		if err = rotator.caSecretController.UpdateCASecretWithRetry(caSecret, rotator.config.retryInterval, 30*time.Second); err != nil {
+			rootCertRotatorLog.Errorf("Failed to rollback CA secret (error: %s).", err.Error())
+			return
+		}
+		if err := rotator.ca.GetCAKeyCertBundle().VerifyAndSetAll(oldCaCert, oldCaPrivateKey, nil, oldCaCert); err != nil {
+			rootCertRotatorLog.Errorf("Failed to rollback CA KeyCertBundle (%s)", err.Error())
+			return
+		}
+		rootCertRotatorLog.Info("Root cert rollback is completed successfully.")
+		return
```

Comment:
I suggest putting this logic in a separate function. rollbackRootCertInSecret().

Reply:
I have refactored this part by using a new method updateRootCertificate() which is used for both rolling forward and rolling backward root certificate.